tests/k8s_utils.go

```diff
@@ -0,0 +1,213 @@
+		if err != nil {
+			return err
+		}
+
+		count := 0
+		for _, pod := range podList.Items {
+			if pod.Status.Phase == expectedPhase {
+				count++
+			}
+		}
+
+		if count == expectedCount {
+			break
+		}
+
+		time.Sleep(5 * time.Second)
+
+		if dumpLog > 6 {
+			fmt.Printf("checking for pod with labelSelector=%s in ns=%s, count=%d expectedCount=%d\n", labelSelector, podNamespace, count, expectedCount)
+			dumpLog = 0
+		}
+		dumpLog++
+	}
+	return nil
+}
+
+func (k *KubeClient) createNamespace(namespace string) error {
+	_, err := k.CoreV1().Namespaces().Get(namespace, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			o := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: namespace,
+				},
+			}
+			_, err = k.CoreV1().Namespaces().Create(o)
+		}
+	}
+	return err
+}
+
+// WaitForNamespaceCleanup wait for cleanup of the given namespace
+func (k *KubeClient) WaitForNamespaceCleanup(ns string) error {
+	dumpLog := 0
+	for {
+		_, err := k.CoreV1().Namespaces().Get(ns, metav1.GetOptions{})
+
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if dumpLog > 6 {
+			fmt.Printf("Waiting for cleanup of namespace %s\n", ns)
+			dumpLog = 0
+		}
+
+		dumpLog++
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (k *KubeClient) destroyNamespace(namespace string) error {
+	err := k.CoreV1().Namespaces().Delete(namespace, &metav1.DeleteOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		return k.WaitForNamespaceCleanup(namespace)
+	}
+	return nil
+}
+
+func (k *KubeClient) waitForPVCBound(pvc, ns string) (corev1.PersistentVolumeClaimPhase, error) {
+	for {
+		o, err := k.CoreV1().
+			PersistentVolumeClaims(ns).
+			Get(pvc, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+
+		if o.Status.Phase == corev1.ClaimLost {
+			return o.Status.Phase, errors.Errorf("PVC %s/%s in lost state", ns, pvc)
+		}
+		if o.Status.Phase == corev1.ClaimBound {
+			return o.Status.Phase, nil
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (k *KubeClient) createPVC(pvc *corev1.PersistentVolumeClaim) error {
+	_, err := k.CoreV1().PersistentVolumeClaims(pvc.Namespace).Create(pvc)
+	if err != nil {
+		if !k8serrors.IsAlreadyExists(err) {
+			return err
+		}
+	}
+
+	_, err = k.waitForPVCBound(pvc.Name, pvc.Namespace)
+	return err
+}
+
+func (k *KubeClient) deletePVC(namespace, pvc string) error {
+	err := k.CoreV1().PersistentVolumeClaims(namespace).Delete(pvc, &metav1.DeleteOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			err = nil
+		}
+	}
+
+	return err
+}
+
+func (k *KubeClient) createDeployment(deployment *appsv1.Deployment) error {
+	_, err := k.AppsV1().Deployments(deployment.Namespace).Create(deployment)
+	if err != nil {
+		return errors.Errorf("Failed to create deployment %s/%s, err=%s", deployment.Namespace, deployment.Name, err)
+	}
```

Comment:
Q: Do we need to handle AlreadyExist case?

Reply:
from test wise not required, but good to add